### PR TITLE
Move the creation of the "fields by JSON name" dictionary to the descriptor

### PR DESCRIPTION
**Pull request by jskeet**
(First of the fast-follow PRs after JSON parsing.)

### CODE REVIEW

**Comment by jskeet** [comment]
(Travis failure appears to be Ruby-related, so irrelevant to this PR.)

File: csharp/src/Google.Protobuf/Reflection/MessageDescriptor.cs [hunk 1]
```diff
@@ -62,6 +62,7 @@ public sealed class MessageDescriptor : DescriptorBase
         private readonly IList<EnumDescriptor> enumTypes;
         private readonly IList<FieldDescriptor> fieldsInDeclarationOrder;
         private readonly IList<FieldDescriptor> fieldsInNumberOrder;
+        private readonly IDictionary<string, FieldDescriptor> jsonFieldMap;
```
Comment (jtattermusch) [inline csharp/src/Google.Protobuf/Reflection/MessageDescriptor.cs]: should we use Lazy (https://msdn.microsoft.com/en-us/library/dd642331(v=vs.110).aspx)?
It's not clear that everyone will use JsonParsing and for mobile (Xamarin), lazy initialization could decrease app startup time. (AFAIK, on Android this has been a concern).

**Comment by jtattermusch** [comment]
Besides the lazy initialization, LGTM.

**Comment by jskeet** [comment]
It's already lazy - we only do any of this when a descriptor is first used. We *could* make the JSON-specific part super-lazy, but I'd rather do that later if we need to. It can be done without an API change, so I'm happy to punt on it for now. Merging.

**Pull request merged**